source/language/scope.rst

```diff
@@ -0,0 +1,318 @@
+
+* Calibration scopes, which are the shared calibration state contained in
+  ``cal`` and ``defcal`` blocks.  The precise rules for these may vary depending
+  on the particular companion calibration grammar that has been loaded for this
+  program.  See :ref:`the OpenPulse specification <openpulse>` for details on
+  how this is handled in the OpenPulse calibration language.  This document will
+  not cover this topic further.
+
+In general, the lifetime of each identifier begins when it is declared, and ends
+at the completion of the scope it was declared in.  The storage space for each
+variable must be allocated for at least the lifetime of the identifier, but it
+is up to individual implementations to define their allocation strategies within
+nested scopes.
+
+In order to be a valid OpenQASM 3.0 program, symbols must be defined before they
+are used; in particular, there is no forward definition of functions and gates,
+and there can be no mutual recursion.
+
+Most regular identifiers of variables can be shadowed in inner scopes, so that
+the identifier temporarily refers to a different variable, but cannot be
+re-declared without entering an inner scope.  The type of the shadowing variable
+does not need to be the same as the type of the variable it is shadowing.
+
+Not all identifiers declared in outer scopes are visible within inner scopes.
+The visibility depends on the type of the variable, and the type of the inner
+scope.  Approximately, ``const`` variables, gates and subroutines are visible
+within *all* inner scopes, while other variables in outer scopes are visible
+within inner control-flow scopes but not gate and function scopes.
+
+The ``include`` statement should be seen as extending the scope it is contained
+within; all variables that are in scope at the time of the ``include`` statement
+are also in scope while the included file is being parsed, and variables defined
+in that file will be available in the containing scope once the inclusion has
+been parsed.  There is no separate namespacing defined in OpenQASM 3.0.
+
+
+.. _scope-global:
+
+Global scope
+------------
+
+The global scope is active when no other scopes are active.  Several OpenQASM
+3.0 statements are only valid in the global scope, such as (non-exhaustive):
+
+* ``gate``, ``def`` and ``defcal`` declarations,
+* ``qubit`` declarations,
+* ``array`` declarations.
+
+A simple example of scoping between a main program file and an included file:
+
+.. code-block:: c
+   :caption: Main program
+
+   OPENQASM 3.0;
+
+   gate h q {
+      U(pi/2, 0, pi) q;
+      // `U` is the single-qubit gate that is implicitly defined in the global
+      // scope of all OpenQASM 3.0 programs, and so is available here.
+      // Similarly `pi` is one of the implicitly defined constants.
+   }
+
+   int i = 100;
+
+   include "my_definitions.qasm";
+
+   // Identifiers 'h', 'my_gate', 'i' and 'SIZE' are defined and in scope.
```

Comment:
Do you mean `j` is in scope after the `include`? What is `SIZE`?

Reply:
Whoops, I changed the example about five times while writing it, and must have got them out of sync.  Thanks for catching it.  Fixed in cf6469f.